modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -23,48 +37,132 @@ import eu.timepit.refined.types.numeric.{NonNegLong, PosLong}
           logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with ordinal=${ordinal.show}")
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger.warn(e)(s"Failure pulling global snapshots") >>
+              Applicative[F].pure(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              snapshotData <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              lastOrdinal = snapshotData.ordinal
+              majorityPeer <- getPeerInTheMajority(lastOrdinal, snapshotData.hash, l0Peers)
+              pulled <- pullSnapshots(majorityPeer, lastStoredOrdinal.next, lastOrdinal)
+            } yield pulled.asRight[LatestSnapshotTuple]
+          }
+        }
+
+      private def pullLatestSnapshotWithMajorityHash(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
+        globalL0ClusterStorage.getPeers
+          .map(nes => Random.shuffle(nes.toSortedSet))
+          .flatMap {
+            _.tailRecM { l0Peers =>
+              l0Peers.headOption
+                .fold(
+                  NoPeersWithMajorityHash.raiseError[F, Either[SortedSet[L0Peer], LatestSnapshotTuple]]
+                ) { l0Peer =>
+                  pullLatestSnapshotFromPeer(l0Peer).attempt.flatMap {
+                    case Left(err) =>
+                      logger
+                        .warn(err)(s"Error pulling latest snapshot from peer ${l0Peer.show}")
+                        .as((l0Peers - l0Peer).asLeft[LatestSnapshotTuple])
+
+                    case Right((hashedSnapshot, info)) =>
+                      findMajorityHash(majorityPeerIds, hashedSnapshot.ordinal).flatMap {
+                        case None =>
+                          logger
+                            .debug(s"No majority hash for latest snapshot of peer ${l0Peer.show}")
+                            .as((l0Peers - l0Peer).asLeft[LatestSnapshotTuple])
+
+                        case Some(hash) if hash === hashedSnapshot.hash =>
```

Comment:
You don't want to check if that snapshot is not to far back in the history? If the majority ordinal would be 1000 it seems ok to accept snapshot with e.g. ordinal 998, 999 or 1000 but maybe 800 is too old? This snapshot at 800 may still be the majority one but far from being latest.